File: packages/php/tests/ClientTest.php
<?php

declare(strict_types=1);

namespace Crawlfox\Tests;

use Crawlfox\Client;
use Crawlfox\CrawlfoxError;
use PHPUnit\Framework\TestCase;

final class ClientTest extends TestCase
{
    public function testRequiresApiKey(): void
    {
        $prev = getenv('CRAWLFOX_API_KEY');
        putenv('CRAWLFOX_API_KEY');
        $this->expectException(\InvalidArgumentException::class);
        try {
            new Client();
        } finally {
            if ($prev !== false) {
                putenv('CRAWLFOX_API_KEY=' . $prev);
            }
        }
    }

    public function testScrapeSendsDocument(): void
    {
        $calls = [];
        $client = $this->client(function (string $method, string $url, array $headers, ?string $body) use (&$calls) {
            $calls[] = compact('method', 'url', 'headers', 'body');
            return [200, json_encode([
                'success' => true,
                'data' => ['markdown' => '# Hello', 'metadata' => ['statusCode' => 200]],
            ])];
        });
        $doc = $client->scrape('https://example.com', [
            'formats' => ['markdown'],
            'country' => 'de',
            'redactPII' => true,
        ]);
        $this->assertTrue($doc['success']);
        $this->assertSame('# Hello', $doc['markdown']);
        $this->assertCount(1, $calls);
        $this->assertSame('POST', $calls[0]['method']);
        $this->assertStringEndsWith('/v1/scrape', $calls[0]['url']);
        $this->assertSame('Bearer cfx_test', $calls[0]['headers']['Authorization']);
        $this->assertStringStartsWith('crawlfox-php/', $calls[0]['headers']['User-Agent']);
        $payload = json_decode($calls[0]['body'], true);
        $this->assertSame('https://example.com', $payload['url']);
        $this->assertSame('de', $payload['country']);
        $this->assertTrue($payload['redactPII']);
    }

    public function testSearchAndBatchAndLogs(): void
    {
        $client = $this->client(function (string $method, string $url, array $headers, ?string $body) {
            if (str_contains($url, '/v1/search')) {
                return [200, json_encode([
                    'success' => true,
                    'data' => ['web' => [['url' => 'https://example.com', 'title' => 'Ex']]],
                    'id' => 'abc',
                ])];
            }
            if (str_contains($url, '/v1/batch')) {
                return [200, json_encode([
                    'success' => true,
                    'count' => 1,
                    'results' => [['success' => true, 'data' => ['markdown' => '# A']]],
                ])];
            }
            if (str_contains($url, '/result')) {
                return [200, json_encode(['ok' => true])];
            }
            if (str_contains($url, '/v1/logs/')) {
                $this->assertSame('GET', $method);
                return [200, json_encode(['id' => 'log1', 'status' => 'ok'])];
            }
            return [404, '{}'];
        });
        $search = $client->search('crawlfox', ['engine' => 'google', 'num' => 5]);
        $this->assertSame('https://example.com', $search['web'][0]['url']);
        $batch = $client->batch(['https://example.org']);
        $this->assertSame('# A', $batch['data'][0]['markdown']);
        $log = $client->getLog('log1');
        $this->assertSame('ok', $log['status']);
        $this->assertSame(['ok' => true], $client->getLogResult('log1'));
    }

    public function testRetriesRetryableThenSucceeds(): void
    {
        $n = 0;
        $client = $this->client(function () use (&$n) {
            $n++;
            if ($n === 1) {
                return [503, json_encode(['message' => 'busy', 'retryable' => true])];
            }
            return [200, json_encode(['success' => true, 'data' => ['markdown' => '# ok']])];
        });
        $doc = $client->scrape('https://example.com');
        $this->assertSame('# ok', $doc['markdown']);
        $this->assertSame(2, $n);
    }

    public function testDoesNotRetryQuota(): void
    {
        $n = 0;
        $client = $this->client(function () use (&$n) {
            $n++;
            return [429, json_encode(['message' => 'quota', 'retryable' => false, 'code' => 'quota'])];
        });
        try {
            $client->scrape('https://example.com');
            $this->fail('expected error');
        } catch (CrawlfoxError $e) {
            $this->assertSame(429, $e->status);
            $this->assertFalse($e->retryable);
            $this->assertSame(1, $n);
        }
    }

    /** @param callable(string, string, array<string, string>, ?string): array{0: int, 1: string} $transport */
    private function client(callable $transport): Client
    {
        return new Client('cfx_test', 'https://api.test', 5, 2, 1, $transport);
    }
}

File: packages/go/client_test.go
package crawlfox

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
)

func TestRequiresAPIKey(t *testing.T) {
	t.Setenv("CRAWLFOX_API_KEY", "")
	os.Unsetenv("CRAWLFOX_API_KEY")
	_, err := New(ClientOptions{})
	if err == nil || !strings.Contains(err.Error(), "API key required") {
		t.Fatalf("expected API key error, got %v", err)
	}
}

func TestScrapeSendsCorrectRequest(t *testing.T) {
	var gotURL, gotAuth, gotBody, gotUA string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotURL = r.URL.Path
		gotAuth = r.Header.Get("Authorization")
		gotUA = r.Header.Get("User-Agent")
		b, _ := io.ReadAll(r.Body)
		gotBody = string(b)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"success":true,"data":{"markdown":"# Hello","metadata":{"statusCode":200}}}`))
	}))
	defer srv.Close()

	client, err := New(ClientOptions{APIKey: "cfx_test", APIURL: srv.URL})
	if err != nil {
		t.Fatal(err)
	}
	res, err := client.Scrape(context.Background(), "https://example.com", &ScrapeOptions{
		Formats:   []string{"markdown"},
		SkipCache: Bool(true),
		Country:   "de",
		RedactPII: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	if res.Markdown != "# Hello" {
		t.Fatalf("unexpected scrape response: %+v", res)
	}
	if gotURL != "/v1/scrape" {
		t.Fatalf("path = %s", gotURL)
	}
	if gotAuth != "Bearer cfx_test" {
		t.Fatalf("auth = %s", gotAuth)
	}
	if !strings.HasPrefix(gotUA, "crawlfox-go/") {
		t.Fatalf("ua = %s", gotUA)
	}
	var body map[string]any
	if err := json.Unmarshal([]byte(gotBody), &body); err != nil {
		t.Fatal(err)
	}
	if body["country"] != "de" || body["redactPII"] != true {
		t.Fatalf("body = %#v", body)
	}
}

func TestSearchSendsQueryBody(t *testing.T) {
	var body map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewDecoder(r.Body).Decode(&body)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"success":true,"data":{"web":[{"url":"https://example.com","title":"Ex"}]}}`))
	}))
	defer srv.Close()

	client, err := New(ClientOptions{APIKey: "cfx_test", APIURL: srv.URL})
	if err != nil {
		t.Fatal(err)
	}
	res, err := client.Search(context.Background(), "crawlfox", &SearchOptions{
		Engine: "google",
		Num:    Int(5),
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(res.Web) == 0 || res.Web[0].URL != "https://example.com" {
		t.Fatalf("unexpected search: %+v", res)
	}
	if body["q"] != "crawlfox" || body["engine"] != "google" {
		t.Fatalf("body = %#v", body)
	}
}

func TestBatch(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"success":true,"count":1,"results":[{"success":true,"data":{"markdown":"# A"}}]}`))
	}))
	defer srv.Close()

	client, err := New(ClientOptions{APIKey: "cfx_test", APIURL: srv.URL})
	if err != nil {
		t.Fatal(err)
	}
	res, err := client.Batch(context.Background(), []string{"https://example.org"}, &ScrapeOptions{
		Formats: []string{"markdown"},
	})
	if err != nil {
		t.Fatal(err)
	}
	if !res.Success || len(res.Data) != 1 || res.Data[0].Markdown != "# A" {
		t.Fatalf("unexpected batch: %+v", res)
	}
}

func TestErrorResponse(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(429)
		_, _ = w.Write([]byte(`{"code":"MONTHLY_QUOTA_EXCEEDED","message":"Quota exceeded","retryable":false}`))
	}))
	defer srv.Close()

	client, err := New(ClientOptions{APIKey: "cfx_test", APIURL: srv.URL})
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.Scrape(context.Background(), "https://example.com", nil)
	cfErr, ok := err.(*Error)
	if !ok {
		t.Fatalf("expected *Error, got %T %v", err, err)
	}
	if cfErr.Status != 429 || cfErr.Code != "MONTHLY_QUOTA_EXCEEDED" {
		t.Fatalf("unexpected error: %+v", cfErr)
	}
}

func TestRejectsEmptyURL(t *testing.T) {
	client, err := New(ClientOptions{APIKey: "cfx_test", APIURL: "http://127.0.0.1:1"})
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.Scrape(context.Background(), "  ", nil)
	if err == nil || !strings.Contains(err.Error(), "url is required") {
		t.Fatalf("got %v", err)
	}
}
